SeedProvider and related stuff

DuncanRuns opened the pull request:
SeedQueue compat is not tested at all yet, wanted to get the code up anyway

Goals of this PR:
- Prevent the need for cheaty code to make something like FSG Mod work
- Allow the changing of provided seed during a session while keeping accurate text in rendering, logging, and debug hud
- Good assertions to prevent hidden mod incompatibilities and misuse of new interfaces

DuncanRuns committed:
SeedProvider and related stuff

DuncanRuns committed:
Simplify loadAtumConfigurations input

DuncanRuns committed:
Remove unused added import

DuncanRuns committed:
Remove unused method

DuncanRuns committed:
Fix seed log

DuncanRuns committed:
assert -> ensureState

DuncanRuns committed:
Suppress client null warning

DuncanRuns committed:
yo

DuncanRuns committed:
Move non static below static

DuncanRuns committed:
Let seed provider provide a waiting screen

DuncanRuns committed:
Add some docs for SeedProvider and add defaults

DuncanRuns committed:
Simplify logic and readability for screen redirect

DuncanRuns committed:
Remove unused import

DuncanRuns committed:
Insane lambda rizz

contariaa commented on src/main/java/me/voidxwalker/autoreset/mixin/config/MoreOptionsDialogMixin.java:
```diff
@@ -37,8 +38,9 @@ public abstract class MoreOptionsDialogMixin implements IMoreOptionsDialog {
-    public void atum$loadAtumConfigurations() {
-        this.seedText = Atum.config.seed;
+    public void atum$loadAtumConfigurations(String seed) {
+        this.seedText = seed == null ? "" : seed;
```
`seed` can't be null since we get it from an `Optional`

contariaa commented on src/main/java/me/voidxwalker/autoreset/mixin/MinecraftClientMixin.java:
```diff
@@ -114,4 +114,10 @@ private boolean clickButton(Screen screen, String... translationKeys) {
         return false;
     }
+
+    @ModifyVariable(method = "startIntegratedServer(Ljava/lang/String;Lnet/minecraft/util/registry/RegistryTracker$Modifiable;Ljava/util/function/Function;Lcom/mojang/datafixers/util/Function4;ZLnet/minecraft/client/MinecraftClient$WorldLoadAction;)V", at = @At("STORE"))
+    private LevelLoadingScreen addSeedToLLS(LevelLoadingScreen levelLoadingScreen, @Local MinecraftClient.IntegratedResourceManager integratedResourceManager) {
+        ((ISeedStringHolder) levelLoadingScreen).atum$setSeedString(((ISeedStringHolder) integratedResourceManager.getSaveProperties().getGeneratorOptions()).atum$getSeedString());
+        return levelLoadingScreen;
+    }
```
The seed string will be `null` when Atum isn't running

contariaa commented on src/main/java/me/voidxwalker/autoreset/mixin/config/CreateWorldScreenMixin.java:
```diff
@@ -94,6 +100,27 @@ private void loadAtumConfigurations(CallbackInfo ci) {
             return;
         }
 
+        if (Atum.isRunning()) {
+            SeedProvider seedProvider = Atum.getSeedProvider();
+            Optional<String> seedOpt = seedProvider.getSeed();
+            if (seedOpt.isPresent()) {
+                seed = seedOpt.get();
+            } else {
+                if (MinecraftClient.getInstance().isOnThread()) {
+                    this.shouldOpenWaitingScreen = true;
+                    return;
+                } else {
+                    while (!seedOpt.isPresent()) {
+                        seedProvider.waitForSeed();
+                        seedOpt = seedProvider.getSeed();
+                    }
+                    seed = seedOpt.get();
+                }
+            }
+        } else {
+            seed = Atum.config.seed; // Set seed for config screen
+        }
+
```
I think this code can just be moved into `modifyAtumCreateWorldScreen` so we don't need two extra fields, you can make an extra method in `IMoreOptionsDialog` for setting the seed string

contariaa reviewed (changes requested):
My main concern is what happens when the reset key is pressed during the waiting screen since `Atum#shouldReset` is set to false before opening the `AtumCreateWorldScreen`. Maybe atum providing the waitingscreen was right, then you could just check that in `Atum#scheduleReset`
I'm also not a big fan of IllegalStateExceptions without reasons, they should give some description of what went wrong, for null checks `Objects.requireNonNull` is fine

DuncanRuns committed:
Waiting screen but only as much as necessary

DuncanRuns committed:
Fix non atum level loading screen

DuncanRuns committed:
Do seed provider stuff later

DuncanRuns committed:
Better ensureState and other checks

DuncanRuns commented:
- AtumWaitingScreen is an abstract Screen class, seed providers will still need to provide an implementation of a waiting screen, but AtumWaitingScreen allows an instanceof check for the hotkey and also allows providing some useful methods that will likely be used
- Code moved to modifyAtumCreateWorldScreen
  - Defensive null check code removed in this refactor as well
- LevelLoadingScreen for non atum loading fixed
- Added `String exceptionMessage` parameter to ensureState, replaced some usages with requireNonNull

contariaa commented on src/main/java/me/voidxwalker/autoreset/mixin/config/CreateWorldScreenMixin.java:
```diff
@@ -137,18 +139,44 @@ private void modifyAtumCreateWorldScreen(CallbackInfo info) {
+                    client.openScreen(Atum.getSeedProvider().getWaitingScreen());
+                    return;
+                } else {
+                    while (!seedOpt.isPresent()) {
+                        seedProvider.waitForSeed();
+                        seedOpt = seedProvider.getSeed();
+                    }
```
I'd rather only wait for the seed once and print a warning/throw an exception if it fails

DuncanRuns replied:
My original reasoning for this was that if SeedQueue ever changed to be able to open multiple AtumCreateWorldScreens, then two waitForSeeds could race and only one would get it and the other should continue to wait.
But if we guarantee that screens can only open one at a time then yeah waiting once would be fine.

contariaa reviewed (changes requested):
`Atum#isSetSeed` should probably return false when a custom seedprovider is registered, its used in chunkcacher so kind of critical
Probably also best to add the AtumWaitingScreen check to `Atum#isBlocking` just in case, prevents it from creating a new world during the waiting screen and instead immediately resets after
Atum should probably ensure it has either stopped running or that a new AtumCreateWorldScreen has been opened when the AtumWaitingScreen is closed, you can probably override `Screen#removed` for that

DuncanRuns committed:
isSetSeed? isConsistentSeed.

DuncanRuns committed:
Throw exception for badly implemented waitForSeed

DuncanRuns committed:
Add AtumWaitingScreen check to isBlocking

DuncanRuns commented:
> Atum should probably ensure it has either stopped running or that a new AtumCreateWorldScreen has been opened when the AtumWaitingScreen is closed, you can probably override `Screen#removed` for that

currentScreen.removed() is called as the first step in MinecraftClient.openScreen(), so I cannot check from removed() if an AtumCreateWorldScreen is being opened.
If I call client.openScreen() from removed() to ensure an AtumCreateWorldScreen then I think it would cause an infinite loop.

DuncanRuns committed:
Suppress unused warnings

contariaa commented:
I think it's fine to just set a flag that a new AtumCreateWorldScreen is created in `AtumWaitingScreen#onSeedFound`, you could even use that to ensure it's only ever called once with a lovely ensureState

DuncanRuns commented:
Ah good idea

DuncanRuns committed:
Ensure proper closing of AtumWaitinScreen

DuncanRuns committed:
formatting

DuncanRuns commented:
Some things to consider:
- Moving AtumCreateWorldScreen to gui package, or moving AtumWaitingScreen out of gui package
- Moving SeedProvider to interfaces package, unless that package is meant for duck interfaces only
- Formatting? I noticed the mixins that I saw don't use multiple lines for the annotation parameters like in SQ so I guess at some point you'd go through everything for formatting if you wanted to anyway.

DuncanRuns committed:
Better names + fix behaviour if shouldCloseOnEsc

DuncanRuns committed:
this.

DuncanRuns commented:
As said in original comment, SeedQueue compatibility wasn't tested and still isn't, this is something I will need to get around to before setting as Ready to review for allowing merging.

DuncanRuns commented:
hmmmm
![image](https://github.com/user-attachments/assets/f227c820-de5f-454c-a665-b26b8aeb4d01)

DuncanRuns commented:
oh I see
this.seedText in CreateWorldScreen is not what is used for setting the seed in actual world gen, seedTextField.getText() is used.
seetTextField.setText(this.seedText) is called from init, which allows changing this.seedText to work if ran before init, such as in constructor. This broke when moving seed provider logic from constructor to init.

DuncanRuns committed:
Fix setting seed

contariaa commented:
> Some things to consider:
> 
> * Moving AtumCreateWorldScreen to gui package, or moving AtumWaitingScreen out of gui package
> 
> * Moving SeedProvider to interfaces package, unless that package is meant for duck interfaces only
> 

I have actually thought about this, I think its best to just move AtumWaitingScreen and SeedProvider into an api package, maybe even with a seedprovider subpackage incase other stuff is ever added

> * Formatting? I noticed the mixins that I saw don't use multiple lines for the annotation parameters like in SQ so I guess at some point you'd go through everything for formatting if you wanted to anyway.

Yea I figured it would be easier if i just did formatting stuff later, doing it through reviews is kind of tedious

contariaa commented:
Oh also you could ensure state in `AtumWaitingScreen#cancelWorldCreation` and `continueWorldCreation` so only one of them is ever called and they are only ever called once

DuncanRuns committed:
Polish AtumWaitingScreen decision + package stuff

DuncanRuns commented:
Something that's missing is some way to cancel seed waiting for SeedQueue (with `waitForSeed()`)
Right now it can just finish waitForSeed and then return an empty optional to force an ISE, but that feels mmmmmmmmm cringe

https://github.com/KingContaria/atum-rewrite/pull/8/files#diff-ae25334a80d0337dfa884e6646c58003073cba72d9d74828cd25b03931c436f9R156

contariaa commented:
You could just return a dummy seed, would probably be easier for me than implementing a way to cancel world creation in seedqueue

DuncanRuns committed:
Fix non atum world creation 2

DuncanRuns committed:
break seedqueue d colon

DuncanRuns committed:
no overriding removed()!!!

contariaa reviewed (approved):
Lgtm :)
